### PR TITLE
Initialize compression state only if they are used.

### PR DESCRIPTION
**Pull request by soheilhy**
Add a lazy initialization path, so that we do not have to
initialize/access the state that's only used when we compress.

Move the frequently access fields to the front so we are more
cache friendly.

### CODE REVIEW

File: src/core/ext/filters/http/message_compress/message_compress_filter.cc [hunk 1]
```diff
@@ -366,6 +380,7 @@ static void start_send_message_batch(void* arg, grpc_error* unused) {
           calld->send_initial_metadata_state == HAS_COMPRESSION_ALGORITHM)) {
     send_message_batch_continue(elem);
   } else {
+    maybe_initialize(elem, calld);
```
Comment (muxi) [inline src/core/ext/filters/http/message_compress/message_compress_filter.cc]: How about putting this at the place where `calld->send_initial_metadata_state` is set to `HAS_COMPRESSION_ALGORITHM`? (currently it is at line 426). In that case you probably can avoid the "maybe" term and `state_initialized`.

File: src/core/ext/filters/http/message_compress/message_compress_filter.cc [hunk 2]
```diff
@@ -60,31 +60,29 @@ struct call_data {
...
+    }
     GRPC_ERROR_UNREF(cancel_error);
   }
 
   grpc_core::CallCombiner* call_combiner;
-  grpc_linked_mdelem compression_algorithm_storage;
-  grpc_linked_mdelem stream_compression_algorithm_storage;
-  grpc_linked_mdelem accept_encoding_storage;
-  grpc_linked_mdelem accept_stream_encoding_storage;
+  grpc_transport_stream_op_batch* send_message_batch = nullptr;
```
Comment (muxi) [inline src/core/ext/filters/http/message_compress/message_compress_filter.cc]: What is the reason to move this field and the *_storage fields around? (I believe you have a good reason but I failed to figure it out; my bad :)

File: src/core/ext/filters/http/message_compress/message_compress_filter.cc (hunk 1 above)
Comment (muxi) [reply src/core/ext/filters/http/message_compress/message_compress_filter.cc]: > Is that common not to have compression algorithm in the initial metadata, and some default compression algorithm?

Ah that is good point! In fact, I am suspecting it is *probably* a bug in the `process_send_initial_metadata`; `has_compression_algorithm` probably should not set to true if the [compression algorithm](https://github.com/grpc/grpc/blob/master/src/core/ext/filters/http/message_compress/message_compress_filter.cc#L166) parsed from the initial metadata is actually "no compression" (@AspirinSJL could you also take a look?). But you are totally right that under this situation, it should not initialize those variables.

File: src/core/ext/filters/http/message_compress/message_compress_filter.cc (hunk 2 above)
Comment (soheilhy) [reply src/core/ext/filters/http/message_compress/message_compress_filter.cc]: Thank you for the feedback!

File: src/core/ext/filters/http/message_compress/message_compress_filter.cc (hunk 1 above)
Comment (soheilhy) [reply src/core/ext/filters/http/message_compress/message_compress_filter.cc]: Yes I agree. We probably shouldn't has_compression to true in the line you mentioned.

Comment (AspirinSJL) [reply src/core/ext/filters/http/message_compress/message_compress_filter.cc]: That seems a bug to me.. `has_compression_algorithm` is always set to true in that function. The code looks strange. I guess we need to stop processing when there is any parsing error.

Comment (AspirinSJL) [reply src/core/ext/filters/http/message_compress/message_compress_filter.cc]: I'm not very familiar with that code, but I can try to change it. If no test fails, then let's change it.

**Comment by soheilhy** [comment]
Thank you @AspirinSJL !

File: src/core/ext/filters/http/message_compress/message_compress_filter.cc (hunk 1 above)
Comment (AspirinSJL) [reply src/core/ext/filters/http/message_compress/message_compress_filter.cc]: Trying to fix this in https://github.com/grpc/grpc/pull/19239.

**Comment by AspirinSJL** [comment]
https://github.com/grpc/grpc/pull/19239 is merged.

**Comment by soheilhy** [comment]
Thank you @AspirinSJL ! I'll pull the new changes and rewrite the patch. Thanks!

File: src/core/ext/filters/http/message_compress/message_compress_filter.cc (hunk 1 above)
Comment (soheilhy) [reply src/core/ext/filters/http/message_compress/message_compress_filter.cc]: Thank you for the fix and sorry for the delay! 

I rebased and did what @muxi suggested.  PTAL. thanks!

File: src/core/ext/filters/http/message_compress/message_compress_filter.cc [hunk 3]
```diff
@@ -225,6 +241,10 @@ static void send_message_batch_continue(grpc_call_element* elem) {
 
 static void finish_send_message(grpc_call_element* elem) {
   call_data* calld = static_cast<call_data*>(elem->call_data);
+
```
Comment (AspirinSJL) [inline src/core/ext/filters/http/message_compress/message_compress_filter.cc]: Please avoid blank lines in functions.

Comment (soheilhy) [reply src/core/ext/filters/http/message_compress/message_compress_filter.cc]: Thank you for catching this. Done.

**Comment by soheilhy** [comment]
Thank you very much! I'll squash and will repush.

**Comment by soheilhy** [comment]
Known issues #17561

**Pull request merged**